Chain of onFailure().withBackOff() may break CompletableFuture invocation

### Context

Java:

```
$ java -version
openjdk version "17.0.8.1" 2023-08-24
OpenJDK Runtime Environment (build 17.0.8.1+1-Ubuntu-0ubuntu120.04)
OpenJDK 64-Bit Server VM (build 17.0.8.1+1-Ubuntu-0ubuntu120.04, mixed mode, sharing)
``` 

The bug appear only if using Mutiny 2.0.0+

### Description

Having the following chained onFailure().withBackOff(), if the first part fail once the second part receive an interrupted Thread:

```java
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;

import io.smallrye.mutiny.Uni;

public class Test {

  public static void main(String[] args) {
    AtomicBoolean test = new AtomicBoolean(false);
    Uni.createFrom().item(() -> {
      if (test.getAndSet(true)) {
        return new Object();
      }
      throw new RuntimeException();
    })
        .onFailure()
        .transform(ex -> {
          ex.printStackTrace(System.err);
          return ex;
        })
        .onFailure()
        .retry()
        .withBackOff(Duration.ofMillis(5), Duration.ofSeconds(5))
        .indefinitely()
        .chain(o -> Uni.createFrom().item(() -> {
          if (Thread.interrupted()) {
            System.out.println("Interrupted!");
            Thread.currentThread().interrupt();
          }
          try {
            return CompletableFuture.supplyAsync(() -> {
              try {
                Thread.sleep(1000);
              } catch (InterruptedException ex) {
                Thread.interrupted();
                throw new RuntimeException(ex);
              }
              return new Object();
            }).get();
          } catch (InterruptedException ex) {
            Thread.interrupted();
            throw new RuntimeException(ex);
          } catch (Exception ex) {
            throw new RuntimeException(ex);
          }
        })
            .onFailure(ex -> false)
            .retry()
            .withBackOff(Duration.ofMillis(100))
            .atMost(1))
    .await()
    .indefinitely();
  }

}
```

The output of the above code is the following:

```
$ java --class-path ~/.m2/repository/io/smallrye/reactive/mutiny/2.0.0/mutiny-2.0.0.jar /tmp/Test.java 
java.lang.RuntimeException
	at Test.lambda$main$0(Test.java:16)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnFailureTransform.subscribe(UniOnFailureTransform.java:31)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:73)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.setOrSwitchUpstream(SwitchableSubscriptionSubscriber.java:205)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.onSubscribe(SwitchableSubscriptionSubscriber.java:107)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
	at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:68)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:74)
	at io.smallrye.mutiny.operators.AbstractMulti.subscribe(AbstractMulti.java:39)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.forward(UniCreateFromPublisher.java:41)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher.subscribe(UniCreateFromPublisher.java:26)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:60)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:65)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:46)
	at Test.main(Test.java:54)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.execute(Main.java:419)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.run(Main.java:192)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.main(Main.java:132)
Interrupted!
Exception in thread "main" java.lang.RuntimeException: java.lang.InterruptedException
	at Test.lambda$main$3(Test.java:44)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:73)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.setOrSwitchUpstream(SwitchableSubscriptionSubscriber.java:205)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.onSubscribe(SwitchableSubscriptionSubscriber.java:107)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
	at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:68)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:74)
	at io.smallrye.mutiny.operators.AbstractMulti.subscribe(AbstractMulti.java:39)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.forward(UniCreateFromPublisher.java:41)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher.subscribe(UniCreateFromPublisher.java:26)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.onNext(UniCreateFromPublisher.java:70)
	at io.smallrye.mutiny.helpers.HalfSerializer.onNext(HalfSerializer.java:30)
	at io.smallrye.mutiny.helpers.StrictMultiSubscriber.onItem(StrictMultiSubscriber.java:84)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.subscription.SerializedSubscriber.onItem(SerializedSubscriber.java:74)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.onItem(MultiRetryWhenOp.java:111)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onItem(UniToMultiPublisher.java:92)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:29)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnFailureTransform.subscribe(UniOnFailureTransform.java:31)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:73)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.setOrSwitchUpstream(SwitchableSubscriptionSubscriber.java:205)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.onSubscribe(SwitchableSubscriptionSubscriber.java:107)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
	at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.resubscribe(MultiRetryWhenOp.java:157)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$TriggerSubscriber.onNext(MultiRetryWhenOp.java:188)
	at io.smallrye.mutiny.helpers.HalfSerializer.onNext(HalfSerializer.java:30)
	at io.smallrye.mutiny.helpers.StrictMultiSubscriber.onItem(StrictMultiSubscriber.java:84)
	at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$ConcatMapMainSubscriber.tryEmit(MultiConcatMapOp.java:188)
	at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$ConcatMapInner.onItem(MultiConcatMapOp.java:282)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onItem(UniToMultiPublisher.java:92)
Caused by: java.lang.InterruptedException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:386)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at Test.lambda$main$3(Test.java:41)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:73)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.setOrSwitchUpstream(SwitchableSubscriptionSubscriber.java:205)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.onSubscribe(SwitchableSubscriptionSubscriber.java:107)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
	at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:68)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp.subscribe(MultiRetryWhenOp.java:74)
	at io.smallrye.mutiny.operators.AbstractMulti.subscribe(AbstractMulti.java:39)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.forward(UniCreateFromPublisher.java:41)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher.subscribe(UniCreateFromPublisher.java:26)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.onNext(UniCreateFromPublisher.java:70)
	at io.smallrye.mutiny.helpers.HalfSerializer.onNext(HalfSerializer.java:30)
	at io.smallrye.mutiny.helpers.StrictMultiSubscriber.onItem(StrictMultiSubscriber.java:84)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.subscription.SerializedSubscriber.onItem(SerializedSubscriber.java:74)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.onItem(MultiRetryWhenOp.java:111)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onItem(UniToMultiPublisher.java:92)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:29)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnFailureTransform.subscribe(UniOnFailureTransform.java:31)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:73)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.setOrSwitchUpstream(SwitchableSubscriptionSubscriber.java:205)
	at io.smallrye.mutiny.subscription.SwitchableSubscriptionSubscriber.onSubscribe(SwitchableSubscriptionSubscriber.java:107)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
	at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.resubscribe(MultiRetryWhenOp.java:157)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$TriggerSubscriber.onNext(MultiRetryWhenOp.java:188)
	at io.smallrye.mutiny.helpers.HalfSerializer.onNext(HalfSerializer.java:30)
	at io.smallrye.mutiny.helpers.StrictMultiSubscriber.onItem(StrictMultiSubscriber.java:84)
	at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$ConcatMapMainSubscriber.tryEmit(MultiConcatMapOp.java:188)
	at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$ConcatMapInner.onItem(MultiConcatMapOp.java:282)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onItem(UniToMultiPublisher.java:92)
	at io.smallrye.mutiny.operators.uni.UniDelayOnItem$UniDelayOnItemProcessor.lambda$onItem$0(UniDelayOnItem.java:53)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Also the main thread does not exit.

### Additional details

We defined the pipeline so that task A have to be retried indefinitely on error with back-off and then task B have to be retried if error apply to some predicate (in this test the predicate never apply but could be a valid one in a real application). The current behavior is that if task A fail once and is then retried a second time with success then the task B receives an interrupted Thread (`Thread.interrupted() == true`). Since task B uses `CompletableFuture.get()` that rely on the `Thread.interrupted()` check to throw an `InterruptedException` the test fails. IMHO the Thread received should not be interrupted if the pipeline continues after an error occurs.

This was not the behavior for previous version up to 1.10.0.

If this finally turns out to not be a Bug would be good to at least document this behavior since CompletableFuture is from the Java SDK.

--- a/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiBufferWithTimeoutOp.java
+++ b/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiBufferWithTimeoutOp.java
@@ -236,7 +236,7 @@ public final class MultiBufferWithTimeoutOp<T> extends AbstractMultiOperator<T, 
         public void onCompletion() {
             if (terminated.compareAndSet(RUNNING, SUCCEED)) {
                 if (task != null) {
-                    task.cancel(true);
+                    task.cancel(false);
                     task = null;
                 }
                 checkedComplete();

--- a/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiDemandPacer.java
+++ b/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiDemandPacer.java
@@ -136,7 +136,7 @@ public class MultiDemandPacer<T> extends AbstractMultiOperator<T, T> {
         @Override
         public void cancel() {
             if (scheduledFuture != null) {
-                scheduledFuture.cancel(true);
+                scheduledFuture.cancel(false);
             }
             super.cancel();
         }

--- a/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiWindowOnDurationOp.java
+++ b/implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiWindowOnDurationOp.java
@@ -267,7 +267,7 @@ public class MultiWindowOnDurationOp<T> extends AbstractMultiOperator<T, Multi<T
                 Future current = container.get();
                 if (current == NONE) {
                     if (task != null) {
-                        task.cancel(true);
+                        task.cancel(false);
                     }
                     return false;
                 }

--- a/implementation/src/main/java/io/smallrye/mutiny/operators/multi/multicast/MultiReferenceCount.java
+++ b/implementation/src/main/java/io/smallrye/mutiny/operators/multi/multicast/MultiReferenceCount.java
@@ -75,7 +75,7 @@ public class MultiReferenceCount<T> extends AbstractMulti<T> implements Multi<T>
         }
 
         ScheduledFuture<?> future = executor.schedule(connection, duration.toMillis(), TimeUnit.MILLISECONDS);
-        connection.setTimer(() -> future.cancel(true));
+        connection.setTimer(() -> future.cancel(false));
     }
 
     void terminated(ConnectableMultiConnection connection) {

--- a/implementation/src/main/java/io/smallrye/mutiny/operators/uni/UniDelayOnItem.java
+++ b/implementation/src/main/java/io/smallrye/mutiny/operators/uni/UniDelayOnItem.java
@@ -41,7 +41,7 @@ public class UniDelayOnItem<T> extends UniOperator<T, T> {
             if (!isCancelled()) {
                 super.cancel();
                 if (scheduledFuture != null) {
-                    scheduledFuture.cancel(true);
+                    scheduledFuture.cancel(false);
                 }
             }
         }

--- a/implementation/src/test/java/io/smallrye/mutiny/groups/UniOnFailureRetryTest.java
+++ b/implementation/src/test/java/io/smallrye/mutiny/groups/UniOnFailureRetryTest.java
@@ -9,6 +9,7 @@ import java.io.IOException;
 import java.time.Duration;
 import java.util.List;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.atomic.AtomicReference;
@@ -557,5 +558,22 @@ public class UniOnFailureRetryTest {
                 .onFailure().retry().withExecutor(null).withBackOff(Duration.ofMillis(100)).atMost(5))
                 .isInstanceOf(IllegalArgumentException.class)
                 .hasMessage("`executor` must not be `null`");
+    }
+
+    @Test
+    void avoidSpuriousInterruptsWithBackoff() {
+        AtomicBoolean flip = new AtomicBoolean();
+        AtomicBoolean interrupted = new AtomicBoolean();
+        String result = Uni.createFrom().item(() -> {
+            if (flip.getAndSet(true)) {
+                return "yolo";
+            }
+            throw new RuntimeException("boom");
+        })
+                .onFailure().retry().withBackOff(Duration.ofMillis(100), Duration.ofSeconds(5)).atMost(1)
+                .onItem().invoke(() -> interrupted.set(Thread.currentThread().isInterrupted()))
+                .await().atMost(Duration.ofSeconds(1));
+        assertThat(result).isEqualTo("yolo");
+        assertThat(interrupted).isFalse();
     }
 }

--- a/implementation/src/test/java/io/smallrye/mutiny/operators/MultiIgnoreTest.java
+++ b/implementation/src/test/java/io/smallrye/mutiny/operators/MultiIgnoreTest.java
@@ -69,7 +69,7 @@ public class MultiIgnoreTest {
                 .onItem().ignoreAsUni().subscribeAsCompletionStage();
 
         assertThat(future).isNotCompleted();
-        future.cancel(true);
+        future.cancel(false);
     }
 
     @Test

--- a/implementation/src/test/java/io/smallrye/mutiny/operators/UniCreateFromFutureTest.java
+++ b/implementation/src/test/java/io/smallrye/mutiny/operators/UniCreateFromFutureTest.java
@@ -371,7 +371,7 @@ public class UniCreateFromFutureTest {
         UniAssertSubscriber<String> subscriber = UniAssertSubscriber.create();
         CompletableFuture<String> cs = new CompletableFuture<>();
         Uni.createFrom().future(cs).subscribe().withSubscriber(subscriber);
-        cs.cancel(true);
+        cs.cancel(false);
         subscriber
                 .awaitFailure()
                 .assertFailedWith(CancellationException.class, null);

